lib/usart/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+//! the calling task to use the USART more directly than if it were a
+//! full-fledged driver of its own.
+
+#![no_std]
+
+#[cfg(any(feature = "stm32h743", feature = "stm32h753"))]
+pub mod stm32h7;
+
+#[cfg(any(feature = "stm32h743", feature = "stm32h753"))]
+use self::stm32h7::Device;
+
+use tinyvec::ArrayVec;
+
+/// Supported baud rates
+#[derive(Debug, Clone, Copy)]
+#[repr(u32)]
+pub enum BaudRate {
+    Rate9600 = 9_600,
+    Rate57600 = 57_600,
+    Rate115200 = 115_200,
+}
+
+/// Handle to an enabled USART device.
+pub struct Usart<const TX_BUF_LEN: usize, const RX_BUF_LEN: usize> {
+    usart: Device,
+    irq_mask: u32,
+    tx_buf: ArrayVec<[u8; TX_BUF_LEN]>,
+    rx_buf: ArrayVec<[u8; RX_BUF_LEN]>,
+    rx_overrun: bool,
+}
+
+/// Errors detected during rx
+pub enum RxError {
+    /// Data has been lost due to rx buffer overrun
+    Overrun,
+}
+
+impl<const TX_BUF_LEN: usize, const RX_BUF_LEN: usize>
+    Usart<TX_BUF_LEN, RX_BUF_LEN>
+{
+    /// Start managing `device`.
+    ///
+    /// Enables the `irq` interrupt; the caller is responsible for calling
+    /// [`Usart::handle_interrupt()`] when that interrupt fires.
+    pub fn new(usart: Device, irq_mask: u32) -> Self {
+        // Turn on our interrupt. We haven't enabled any interrupt sources at
+        // the USART side yet, but we will momentarily.
+        userlib::sys_irq_control(irq_mask, true);
+
+        // Enable RX interrupts from the USART side.
+        usart.enable_rx_interrupts();
+
+        Self {
+            usart,
+            irq_mask,
+            tx_buf: ArrayVec::new(),
+            rx_buf: ArrayVec::new(),
+            rx_overrun: false,
+        }
+    }
+
+    /// Handle (by stepping tx/rx if possible) and then reenable the USART
+    /// interrupt.
+    pub fn handle_interrupt(&mut self) {
+        // See if we have any data to transmit.
+        if let Some(&byte) = self.tx_buf.get(0) {
+            // Write it to the transmitter, if possible.
+            if self.usart.try_write_tx(byte) {
+                // TODO? `remove()` shifts all remaining tx data down; we could
+                // use somehting more ringbuffer-like if this is too expensive.
+                self.tx_buf.remove(0);
+                if self.tx_buf.is_empty() {
+                    self.usart.disable_tx_interrupts();
+                }
+            }
+        }
+
+        // See if any data has come in.
+        if let Some(byte) = self.usart.try_read_rx() {
+            if self.rx_buf.try_push(byte).is_some() {
+                // self.rx_buf is full; we'll treat this the same as a hardware
+                // overrun: discard the byte and return an error later
+                self.rx_overrun = true;
+            }
+        }
+
+        // reenable interrupt
+        userlib::sys_irq_control(self.irq_mask, true);
+    }
+
+    /// Get mutable access to the tx/rx buffers held by `self`.
+    pub fn buffers(
```

Comment:
This API is a little odd, could you have read and write functions directly on the Usart? It is rare in practice for the read and write sides to be handled from different code (you may have partially gotten this from the original stm32h7 usart demo task, which is a bit of an anomaly).

Reply:
Agreed on the "a little odd" part. I originally had read/write directly on the Usart, but ran into lifetime issues trying to read the contents of the rx buffer in a loop that also pushed things into tx. I'll roll this in with the work to use the FIFO and get rid of the buffering altogether.